### ISSUE TITLE
consider constructor delegation when determining definitely specified 

### ISSUE THREAD

**Issue by jvasileff** [ceylon/ceylon-spec#1338]
The code below is safe, but disallowed. It might be nice to support this sort of thing, so that you can organize constructors (possibly non-`shared`) in a way that allows the use of methods in constructors.

```ceylon
class ItIsSpecified {
    variable Integer i;

    shared new() {
        i = 1;
    }

    void increment() => i++;
    // error: Not definitely initialized: i

    shared new Two() extends ItIsSpecified() {
        increment();
    }
}
```

**Comment by gavinking** [comment]
I'm against trying to do anything in this area. That would require analysis of the code to prove that the _only_ thing that calls `increment()` is `Two()` and that in full generality is quite an open-ended sort of flow analysis and not the kind of reasoning the typechecker ever attempts. The sort of reasoning we do so far is a purely lexical sort of thing that can be done in a single pass of the code. And even with this braindead system it's already *much* too slow, by the way. As much as 20% of the time in the typechecker is spent analyzing definite assignment.

**Comment by jvasileff** [comment]
Yeah, I know the nature of this analysis is a bit different than what currently exists, so I perfectly understand if you don't want to go down this path.

On the other hand, this may even be an optimization!

1. `increment()` is safe and currently allowed in the initializer section, absent subsequent constructors
2. It is impossible for a constructor `C` that delegates to any preceding constructor to contribute to the unsafety of any code preceding `C` WRT accessing potentially uninitialized declarations.

So, from a top-down *reading* of the code, it seems trivial to prove that `Two` does not make `increment` unsafe. In fact, `Two`'s very first act is to delegate (goto) to an entry point for which `increment` is already known to be safe. (This of course relies on other, existing restrictions, like `increment` being disallowed in `Two`'s `extends` clause.)
